### PR TITLE
Preparatory refactors/fixes/cleanups for GCR

### PR DESCRIPTION
**Pull request by briantigerchow**
@whyrusleeping @jbenet 

These are preparatory modifications for GCR. Handling these modifications upfront will cut the CR for grandcentral routing down by a factor of 2.

### CODE REVIEW

File: cmd/ipfs/daemon.go [hunk 1]
```diff
@@ -91,16 +92,15 @@ func daemonFunc(req cmds.Request, res cmds.Response) {
 		res.SetError(debugerror.Errorf("Couldn't obtain lock. Is another daemon already running?"), cmds.ErrNormal)
 		return
 	}
-	defer repo.Close()
 
 	// OK!!! Now we're ready to construct the node.
 	// make sure we construct an online node.
-	ctx.Online = true
-	node, err := ctx.GetNode()
+	node, err := core.NewIPFSNode(ctx.Context, core.Online(repo))
```
Comment (jbenet) [inline cmd/ipfs/daemon.go]: :+1: yaaay

File: exchange/bitswap/bitswap.go [hunk 2]
```diff
@@ -52,28 +53,47 @@ var (
 func New(parent context.Context, p peer.ID, network bsnet.BitSwapNetwork,
 	bstore blockstore.Blockstore, nice bool) exchange.Interface {
 
+	// important to use provided parent context (since it may include important
+	// loggable data). It's probably not a good idea to allow bitswap to be
+	// coupled to the concerns of the IPFS daemon in this way.
+	//
+	// FIXME(btc) Now that bitswap manages itself using a process, it probably
+	// shouldn't accept a context anymore. Clients should probably use Close()
+	// exclusively. We should probably find another way to share logging data
```
Comment (jbenet) [inline exchange/bitswap/bitswap.go]: agreed. creating new contexts (which we do in various places) looses the logging information.

This lack of context awareness is a drawback in the `goprocess` lib. I can make it be context aware (i.e. make processes have a `Context() context.Context` but i dont know this is the right thing to do (it moves it back to the CtxCloser world it got out of). what would make it easier on the logging/user?

File: p2p/protocol/mux.go [hunk 3]
```diff
@@ -24,11 +24,11 @@ type streamHandlerMap map[ID]inet.StreamHandler
 // It contains the handlers for each protocol accepted.
 // It dispatches handlers for streams opened by remote peers.
 type Mux struct {
-	// defaultHandler handles unknown protocols. Callers modify at your own risk.
-	defaultHandler inet.StreamHandler
 
 	lock     sync.RWMutex
 	handlers streamHandlerMap
+	// defaultHandler handles unknown protocols. Callers modify at your own risk.
```
Comment (jbenet) [inline p2p/protocol/mux.go]: i think the comment can be removed now?

File: repo/fsrepo/component/datastore.go [hunk 4]
```diff
@@ -39,9 +38,6 @@ func init() {
 func InitDatastoreComponent(dspath string, conf *config.Config) error {
 	// The actual datastore contents are initialized lazily when Opened.
 	// During Init, we merely check that the directory is writeable.
-	if !filepath.IsAbs(dspath) {
-		return debugerror.New("datastore filepath must be absolute") // during initialization (this isn't persisted)
-	}
```
Comment (jbenet) [inline repo/fsrepo/component/datastore.go]: :+1: yaaay

File: test/integration/three_legged_cat_test.go [hunk 5]
```diff
@@ -23,7 +23,7 @@ func TestThreeLeggedCat1KBInstantaneous(t *testing.T) {
 		RoutingLatency:    0,
 		BlockstoreLatency: 0,
 	}
-	if err := RunThreeLeggedCat(RandomBytes(1*unit.KB), conf); err != nil {
+	if err := RunThreeLeggedCat(RandomBytes(100*unit.MB), conf); err != nil {
```
Comment (jbenet) [inline test/integration/three_legged_cat_test.go]: i think tis is an error. this is supposed to be `TestThreeLeggedCat1KBInstantaneous` ?

I notice that all the rest are 1KB (even the one that says its 100MB): https://github.com/jbenet/go-ipfs/blob/refactors/prep-for-gcr/test/integration/three_legged_cat_test.go#L58

File: thirdparty/eventlog/log.go [hunk 6]
```diff
@@ -111,13 +110,31 @@ func (el *eventLogger) Event(ctx context.Context, event string, metadata ...Logg
 	e.Log() // TODO replace this when leveled-logs have been implemented
 }
 
-type doneCloserFunc func()
+type EventInProgress struct {
+	loggables []Loggable
+	doneFunc  func([]Loggable)
+}
```
Comment (jbenet) [inline thirdparty/eventlog/log.go]: :+1: i like this

**Comment by jbenet** [comment]
several comments up there, but otherwise LGTM.

**Comment by briantigerchow** [comment]
addressed comments. will merge if tests pass

**Pull request merged**